### PR TITLE
Fetch new validation data for each calc_loss

### PR DESCRIPTION
**Pull request by Sopel97**
This PR makes `calc_loss` use a new validation set each time it's called (one can now limit the amount of validation data used per `calc_loss` call, so this makes sense). Since we're using relatively small validation sets for each calc_loss this should, in principle at least, make validation loss more informative when observed over multiple validation steps.

I'd like to get some feedback from @noobpwnftw on this change before it's merged.

I'd also like to get some feedback on how to handle the case when we couldn't fetch the requested amount of the data for validation. This might happen in extreme cases. Currently I made it continue with a warning as the training can proceed even without validation data.

### CODE REVIEW

**Comment by NightlyKing** [comment]
As I already said in discord: the loss now really jumps around like crazy. My idea was to only slowly introduce new positions to the validation set to get a more stable loss output which will help with automatic LR drops.
